Add steps to configure a PGP key for agent upgrade (#984)
elastic/ingest-docs: docs/en/ingest-management/fleet/air-gapped.asciidoc

xpack.fleet.isAirGapped: true
----

[discrete]
[[air-gapped-pgp-fleet]]
== Configure {agents} to download a PGP/GPG key from {fleet-server}

Starting from version 8.9.0, when {agent} tries to perform an upgrade, it first verifies the binary signature with the key bundled in the agent. This process has a backup mechanism that will use the key coming from `https://artifacts.elastic.co/GPG-KEY-elastic-agent` instead of the one it already has.

> In an air-gapped environment, the agent won't be able to download the remote key and therefore cannot be upgraded. For versions 8.9.0 to 8.10.3, you can resolve this problem following the steps described in the associated link:https://www.elastic.co/guide/en/fleet/8.9/release-notes-8.9.0.html#known-issues-8.9.0[known issue] documentation.

Reviewer: ```suggestion
In an air-gapped environment, an {agent} which doesn't have access PGP/GPG key from `https://artifacts.elastic.co/GPG-KEY-elastic-agent` would fail to be upgraded.

For versions 8.9.0 to 8.10.3, you can resolve this problem following the steps described in the associated link:https://www.elastic.co/guide/en/fleet/8.9/release-notes-8.9.0.html#known-issues-8.9.0[known issue] documentation.
```

Author: Thanks! Added in [f698c5e](https://github.com/elastic/ingest-docs/pull/984/commits/f698c5e0a0d5281f279d95865d0c3dd62019841e)

> Starting in version 8.10.4, you can resolve this problem by configuring {agents} to download the PGP/GPG key from {fleet-server}.

Reviewer: ```suggestion
Starting in version 8.10.4, {agent} will:
- verify the binary signature with the key bundled in the agent - if it doesn't pass
- download the PGP/GPG key from `https://artifacts.elastic.co/GPG-KEY-elastic-agent` and verify it - if it doesn't pass
- download the PGP/GPG key from {fleet-server} and verify it
```

Author: Thanks! Added in [f698c5e](https://github.com/elastic/ingest-docs/pull/984/commits/f698c5e0a0d5281f279d95865d0c3dd62019841e)

By default, {fleet-server} serves {agents} with the key located in `FLEETSERVER_BINARY_DIR/elastic-agent-upgrade-keys/default.pgp`.

> The key is served through the endpoint `FLEETSERVER_BINARY_DIR/elastic-agent-upgrade-keys/default.pgp`.

Reviewer: ```suggestion
The key is served through the Fleet Server endpoint `GET /api/agents/upgrades/{major}.{minor}.{patch}/pgp-public-key`.
```

Author: Thanks! Added in [f698c5e](https://github.com/elastic/ingest-docs/pull/984/commits/f698c5e0a0d5281f279d95865d0c3dd62019841e)

> Accessing the endpoint requires a valid API key and the endpoint is rate limited.

Reviewer: ```suggestion
Accessing the endpoint requires a valid API key and the endpoint is rate limited.
```

I am not sure we're protecting the key via API key - can we confirm?

> If there isn't a `default.pgp` key in the `FLEETSERVER_BINARY_DIR/elastic-agent-upgrade-keys/default.pgp` directory, {fleet-server} instead will attempt to retrieve a PGP/GPG key from the URL that you can specify with the `server.pgp.upstream_url` setting.

Reviewer: ```suggestion
If there isn't a `default.pgp` key in the `FLEETSERVER_BINARY_DIR/elastic-agent-upgrade-keys/default.pgp` directory, {fleet-server} instead will attempt to retrieve a PGP/GPG key from the URL that you can specify with the `server.pgp.upstream_url` setting.
```

Few points to clarify:

- Is the `default.pgp` bundled with Fleet Server?

If it's bundled, air-gapped users are good to go with it.

- Is the `default.pgp` downloaded lazily from `https://artifacts.elastic.co/GPG-KEY-elastic-agent` by default?

If it is the case, then air-gapped users can grant access to it (HTTP Proxy or open the firewall).

- Is the `default.php` downloaded lazily from the URL provided via `server.pgp.upstream_url` setting?

If it is the case, then air-gapped users can specify a URL so that it will be downloaded from there.

Reviewer 2: It's a lazy download from `server.pgp.upstream_url`

Reviewer 2: To be very clear: 
> * Is the `default.pgp` bundled with Fleet Server?

No, on install this file does not exist; but a user can add it manually write the file on the instance if there is 0 internet access in order to distribute it to the agents

> * Is the `default.pgp` downloaded lazily from `https://artifacts.elastic.co/GPG-KEY-elastic-agent` by default?

Yes, the default value of `server.pgp.upstream_url` is `https://artifacts.elastic.co/GPG-KEY-elastic-agent`; if `default.pgp` is not on-disk when the key is requested, fleet-server will download the file from the upstream URL, save it to disk and serve it

> * Is the `default.php` downloaded lazily from the URL provided via `server.pgp.upstream_url` setting?

yes, see above 😄

Author: @michel-laterman I'm sorry to bring this one up again (I lost track of it and am doing some housekeeping now). 

Is there anything I need to clarify in the text above, or do you think this is okay to merge as is?

Reviewer 2: I don't think so

xpack.fleet.isAirGapped: true
----

[discrete]
[[air-gapped-pgp-fleet]]
== Configure {agents} to download a PGP/GPG key from {fleet-server}

Starting from version 8.9.0, when {agent} tries to perform an upgrade, it first verifies the binary signature with the key bundled in the agent. This process has a backup mechanism that will use the key coming from `https://artifacts.elastic.co/GPG-KEY-elastic-agent` instead of the one it already has.

In an air-gapped environment, the agent won't be able to download the remote key and therefore cannot be upgraded. For versions 8.9.0 to 8.10.3, you can resolve this problem following the steps described in the associated link:https://www.elastic.co/guide/en/fleet/8.9/release-notes-8.9.0.html#known-issues-8.9.0[known issue] documentation.

> Starting in version 8.10.4, you can resolve this problem by configuring {agents} to download the PGP/GPG key from a custom URL specified in {fleet-server}:

Reviewer 2: I think my reading comprehension on my 1st pass wasn't great.

Just to be clear fleet-server serves PGP keys at `GET /api/agents/upgrades/{major}.{minor}.{patch}/pgp-public-key`
It will serve the key in `FLEETSERVER_BINARY_DIR/elastic-agent-upgrade-keys/default.pgp` by default, if there is no `default.pgp` key in the dir it will look upstream to retrieve a key; `server.pgp.upstream_url`  can be used to configure to a custom URL for fleet-server to use.

IIRC agents will try to get the upstream key, then the fleet-server endpoint (@michalpristas please confirm), they do not use the url we add to fleet-server config

Author: Thanks @michel-laterman! That's really clear. Here's my attempt to fix things up:

---

![Screenshot 2024-07-30 at 12 37 29 PM](https://github.com/user-attachments/assets/7d541046-db25-495c-9326-95dd30e9e3ee)

For versions 8.9.0 to 8.10.3, you can resolve this problem following the steps described in the associated link:https://www.elastic.co/guide/en/fleet/8.9/release-notes-8.9.0.html#known-issues-8.9.0[known issue] documentation.

Starting in version 8.10.4, you can resolve this problem by configuring {agents} to download the PGP/GPG key from {fleet-server}.

Starting in version 8.10.4, {agent} will:

. Verify the binary signature with the key bundled in the agent.
. If the verification doesn't pass, the agent will download the PGP/GPG key from `https://artifacts.elastic.co/GPG-KEY-elastic-agent` and verify it.
. If that verification doesn't pass, the agent will download the PGP/GPG key from {fleet-server} and verify it.
. If that verification doesn't pass, the upgrade is blocked.

By default, {fleet-server} serves {agents} with the key located in `FLEETSERVER_BINARY_DIR/elastic-agent-upgrade-keys/default.pgp`.
The key is served through the {fleet-server} endpoint `GET /api/agents/upgrades/{major}.{minor}.{patch}/pgp-public-key`.

> If there isn't a `default.pgp` key in the `FLEETSERVER_BINARY_DIR/elastic-agent-upgrade-keys/default.pgp` directory, {fleet-server} instead will attempt to retrieve a PGP/GPG key from the URL that you can specify with the `server.pgp.upstream_url` setting.

Reviewer: Can we add a statement to tell users can download the file manually instead of waiting for Fleet Server to lazily download the file? This might be useful for Fleet Servers which do not have access to the public URL.

Author: Thanks! I think this is addressed by your comment below, and I've committed it.

To set a custom URL for {fleet-server} to access a PGP/GPG key and make it available to {agents}:

. In {kib}, go to *Management > {fleet} > Agent policies*.
. Select a policy for the agents that you want to upgrade.
. On the policy page, in the **Actions** menu for the {fleet-server} integration, select **Edit integration**.
. In the {fleet-server} settings section expand **Change defaults** and **Advanced options**.
. In the **Custom fleet-server configurations** field, add the setting `server.pgp.upstream_url` with the URL where the PGP/GPG key can be accessed. For example:

[source,yaml]
----

> server.pgp.upstream_url: <my-web-server-url>

Reviewer: It must be the full url or just the host?
E.g. `http://my-web:8080/default.pgp` or `http://my-web:8080/`?

Author: @michel-laterman Can you help with Luca's question?

Author: Updated to:

```
In the **Custom fleet-server configurations** field, add the setting `server.pgp.upstream_url` 
with the full URL where the PGP/GPG key can be accessed. For example:
[source,yaml]
----
server.pgp.upstream_url: <http://my-web-server:8080/default.pgp>
```

----

The setting `server.pgp.upstream_url` must point to a web server hosting the PGP/GPG key, which must be reachable by the host where {fleet-server} is installed.

Note that:

 * `server.pgp.upstream_url` may be specified as an `http` endpoint (instead of `https`).

>  * For an `https` endpoint, the CA must be trusted by {fleet-server}.

Reviewer: Is there a setting where one can provide the CA to trust the TLS Certificate of the Fleet server?

Author: @michel-laterman Can you help with Luca's question?

Reviewer 2: I believe that it would be the standard CA options

Reviewer: This is Fleet Server integration connecting to `server.pgp.upstream_url`. 
Would it use the `--certificate-authorities` (the global one for EA)?

Author: @michel-laterman Tagging you just in case you haven't seen Luca's question. Thanks!

Reviewer 2: Yes, it should

Author: Thanks! I've changed the text to:

>For an `https` endpoint, the CA for {fleet-server} to connect to `server.pgp.upstream_url` must be trusted by {fleet-server} using the `--certificate-authorities` setting that is used globally for {agent}.

By default, {fleet-server} serves {agents} with the key located in `FLEETSERVER_BINARY_DIR/elastic-agent-upgrade-keys/default.pgp`.
The key is served through the {fleet-server} endpoint `GET /api/agents/upgrades/{major}.{minor}.{patch}/pgp-public-key`.

If there isn't a `default.pgp` key in the `FLEETSERVER_BINARY_DIR/elastic-agent-upgrade-keys/default.pgp` directory, {fleet-server} instead will attempt to retrieve a PGP/GPG key from the URL that you can specify with the `server.pgp.upstream_url` setting.

You can prevent {fleet} from downloading the PGP/GPG key from `server.pgp.upstream_url` by manually downloading it from `https://artifacts.elastic.co/GPG-KEY-elastic-agent` and storing it at  `FLEETSERVER_BINARY_DIR/elastic-agent-upgrade-keys/default.pgp`.

To set a custom URL for {fleet-server} to access a PGP/GPG key and make it available to {agents}:

. In {kib}, go to *Management > {fleet} > Agent policies*.
. Select a policy for the agents that you want to upgrade.
. On the policy page, in the **Actions** menu for the {fleet-server} integration, select **Edit integration**.
. In the {fleet-server} settings section expand **Change defaults** and **Advanced options**.

> . In the **Custom fleet-server configurations** field, add the setting `server.pgp.upstream_url` with the URL where the PGP/GPG key can be accessed. For example:
> 
> [source,yaml]
> ----
> server.pgp.upstream_url: <my-web-server-url>

Author: ```suggestion
. In the **Custom fleet-server configurations** field, add the setting `server.pgp.upstream_url` with the full URL where the PGP/GPG key can be accessed. For example:

[source,yaml]
----
server.pgp.upstream_url: <http://my-web-server:8080/default.pgp>
```

----

The setting `server.pgp.upstream_url` must point to a web server hosting the PGP/GPG key, which must be reachable by the host where {fleet-server} is installed.

Note that:

 * `server.pgp.upstream_url` may be specified as an `http` endpoint (instead of `https`).

>  * For an `https` endpoint, the CA must be trusted by {fleet-server}.

Author: ```suggestion
 * For an `https` endpoint, the CA for {fleet-server} to connect to `server.pgp.upstream_url` must be trusted by {fleet-server} using the `--certificate-authorities` setting that is used globally for {agent}.
```
